File: lobe/src/services/ImageTaskLoggerService.ts
import { fetch, get, post, putJson } from '../utils/fetch';

export interface ImageTaskLogger {
  id: string;
  taskId: string;
  taskType: number;
  taskStatus: number;
  prompt: string;
  imageUrls?: string;
  quota: number;
  modelName: string;
  tokenName?: string;
  userName?: string;
  userId?: string;
  channelId?: string;
  channelName?: string;
  totalTime: number;
  ip?: string;
  userAgent?: string;
  organizationId?: string;
  url?: string;
  isSuccess: boolean;
  errorMessage?: string;
  taskParameters?: string;
  progress: number;
  taskCreatedAt?: string;
  taskCompletedAt?: string;
  createdAt: string;
  updatedAt?: string;
}

export interface ImageTaskLoggerRequest {
  page: number;
  pageSize: number;
  taskType?: number;
  taskStatus?: number;
  model?: string;
  startTime?: string;
  endTime?: string;
  keyword?: string;
  userId?: string;
}

export interface ImageTaskStatistics {
  taskTypes: Array<{
    taskType: string;
    total: number;
    completed: number;
    failed: number;
    processing: number;
    totalQuota: number;
  }>;
  totalTasks: number;
  totalQuota: number;
  successRate: number;
}

// 获取图片任务日志列表
export function getImageTaskLoggers(params: ImageTaskLoggerRequest) {
  const queryParams = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      queryParams.append(key, value.toString());
    }
  });
  return get(`/api/v1/image-tasks?${queryParams.toString()}`);
}

// 根据TaskId获取任务详情
export function getImageTaskByTaskId(taskId: string) {
  return get(`/api/v1/image-tasks/${taskId}`);
}

// 获取图片任务统计信息
export function getImageTaskStatistics(startTime?: string, endTime?: string) {
  const queryParams = new URLSearchParams();
  if (startTime) queryParams.append('startTime', startTime);
  if (endTime) queryParams.append('endTime', endTime);
  return get(`/api/v1/image-tasks/statistics?${queryParams.toString()}`);
}

// 更新任务状态
export function updateImageTaskStatus(taskId: string, data: {
  status: number;
  progress?: number;
  imageUrls?: string[];
  errorMessage?: string;
}) {
  return fetch(`/api/v1/image-tasks/${taskId}/status`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data)
  });
}

// 导出图片任务日志
export function exportImageTasks(params: Omit<ImageTaskLoggerRequest, 'page' | 'pageSize'>) {
  const queryParams = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      queryParams.append(key, value.toString());
    }
  });
  return get(`/api/v1/image-tasks/export?${queryParams.toString()}`);
}

// 获取任务类型列表
export function getTaskTypes() {
  return get('/api/v1/image-tasks/task-types');
}

// 获取任务状态列表
export function getTaskStatuses() {
  return get('/api/v1/image-tasks/task-statuses');
}

// 任务类型映射
export const TaskTypeMap: Record<number, string> = {
  1: 'Imagine',
  2: 'Blend',
  3: 'Describe',
  4: 'Shorten',
  5: 'Modal'
};

// 任务状态映射
export const TaskStatusMap: Record<number, string> = {
  1: '已提交',
  2: '处理中',
  3: '已完成',
  4: '失败',
  5: '已取消'
};

// 任务状态颜色映射
export const TaskStatusColorMap: Record<number, string> = {
  1: 'default',
  2: 'secondary',
  3: 'default',
  4: 'destructive',
  5: 'outline'
};

// 事件监听器类型
export type ImageTaskEventListener = (task: ImageTaskLogger) => void;

// 事件管理器
class ImageTaskEventManager {
  private listeners: Map<string, ImageTaskEventListener[]> = new Map();
  private ws: WebSocket | null = null;
  private reconnectTimer: NodeJS.Timeout | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;

  // 连接 WebSocket
  connect() {
    if (this.ws?.readyState === WebSocket.OPEN) return;

    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
    const wsUrl = `${protocol}//${window.location.host}/ws/image-tasks`;

    this.ws = new WebSocket(wsUrl);

    this.ws.onopen = () => {
      console.log('Image task WebSocket connected');
      this.reconnectAttempts = 0;
    };

    this.ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        this.handleEvent(data);
      } catch (error) {
        console.error('Failed to parse WebSocket message:', error);
      }
    };

    this.ws.onclose = () => {
      console.log('Image task WebSocket disconnected');
      this.scheduleReconnect();
    };

    this.ws.onerror = (error) => {
      console.error('Image task WebSocket error:', error);
    };
  }

  // 断线重连
  private scheduleReconnect() {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('Max reconnection attempts reached');
      return;
    }

    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000);
    this.reconnectTimer = setTimeout(() => {
      this.reconnectAttempts++;
      this.connect();
    }, delay);
  }

  // 处理事件
  private handleEvent(data: any) {
    const { type, taskId, task } = data;
    const listeners = this.listeners.get(type) || [];
    listeners.forEach(listener => listener(task));

    // 触发特定任务 ID 的监听器
    const taskListeners = this.listeners.get(`task:${taskId}`) || [];
    taskListeners.forEach(listener => listener(task));
  }

  // 订阅事件
  subscribe(eventType: string, listener: ImageTaskEventListener) {
    if (!this.listeners.has(eventType)) {
      this.listeners.set(eventType, []);
    }
    this.listeners.get(eventType)!.push(listener);

    // 确保 WebSocket 连接
    this.connect();

    return () => this.unsubscribe(eventType, listener);
  }

  // 取消订阅
  unsubscribe(eventType: string, listener: ImageTaskEventListener) {
    const listeners = this.listeners.get(eventType);
    if (listeners) {
      const index = listeners.indexOf(listener);
      if (index > -1) {
        listeners.splice(index, 1);
      }
    }
  }

  // 订阅特定任务
  subscribeToTask(taskId: string, listener: ImageTaskEventListener) {
    return this.subscribe(`task:${taskId}`, listener);
  }

  // 订阅所有任务状态变化
  subscribeToStatusChanges(listener: ImageTaskEventListener) {
    return this.subscribe('status_change', listener);
  }

  // 订阅任务完成事件
  subscribeToTaskCompletion(listener: ImageTaskEventListener) {
    return this.subscribe('task_completed', listener);
  }

  // 订阅任务失败事件
  subscribeToTaskFailure(listener: ImageTaskEventListener) {
    return this.subscribe('task_failed', listener);
  }

  // 断开连接
  disconnect() {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }

    this.listeners.clear();
  }
}

// 全局事件管理器实例
export const imageTaskEventManager = new ImageTaskEventManager();

File: src/Thor.Service/Service/ImageTaskLoggerService.cs
using System.Diagnostics;
using System.Text.Json;
using Thor.BuildingBlocks.Event;
using Thor.Domain.Images;
using Thor.Domain.Shared;
using Thor.Infrastructure;

namespace Thor.Service.Service;

/// <summary>
/// 图片任务日志服务
/// </summary>
public sealed class ImageTaskLoggerService(
    IServiceProvider serviceProvider,
    IEventBus<ImageTaskLogger> eventBus,
    IServiceCache serviceCache)
    : ApplicationService(serviceProvider)
{
    /// <summary>
    /// 创建图片任务日志
    /// </summary>
    public async ValueTask CreateAsync(ImageTaskLogger logger)
    {
        logger.CreatedAt = DateTime.Now;
        await eventBus.PublishAsync(logger);
    }

    /// <summary>
    /// 创建图片任务提交日志
    /// </summary>
    public async ValueTask CreateTaskAsync(
        string taskId,
        ThorImageTaskType taskType,
        string prompt,
        string model,
        long quota,
        string? tokenName,
        string? userName,
        string? userId,
        string? channelId,
        string? channelName,
        string ip,
        string userAgent,
        string url,
        int totalTime,
        string? organizationId = null,
        bool isSuccess = true,
        string? errorMessage = null,
        object? taskParameters = null)
    {
        using var activity = Activity.Current?.Source.StartActivity("创建图片任务日志");

        activity?.SetTag("TaskId", taskId);
        activity?.SetTag("TaskType", taskType.ToString());
        activity?.SetTag("Model", model);

        var logger = new ImageTaskLogger
        {
            TaskId = taskId,
            TaskType = taskType,
            TaskStatus = isSuccess ? ThorImageTaskStatus.Submitted : ThorImageTaskStatus.Failed,
            Prompt = prompt,
            ModelName = model,
            Quota = quota,
            TokenName = tokenName,
            UserName = userName,
            UserId = userId,
            ChannelId = channelId,
            ChannelName = channelName,
            TotalTime = totalTime,
            IP = ip,
            UserAgent = userAgent,
            OrganizationId = organizationId,
            Url = url,
            IsSuccess = isSuccess,
            ErrorMessage = errorMessage,
            TaskParameters = taskParameters != null ? JsonSerializer.Serialize(taskParameters) : null,
            TaskCreatedAt = DateTime.Now
        };

        await CreateAsync(logger);
    }

    /// <summary>
    /// 更新任务状态
    /// </summary>
    public async ValueTask UpdateTaskStatusAsync(
        string taskId,
        ThorImageTaskStatus status,
        int progress = 0,
        string[]? imageUrls = null,
        string? errorMessage = null)
    {
        var logger = await LoggerDbContext.ImageTaskLoggers
            .FirstOrDefaultAsync(x => x.TaskId == taskId);

        if (logger == null) return;

        logger.TaskStatus = status;
        logger.Progress = progress;
        logger.ErrorMessage = errorMessage;

        if (imageUrls?.Length > 0)
        {
            logger.ImageUrls = JsonSerializer.Serialize(imageUrls);
        }

        if (status == ThorImageTaskStatus.Completed || status == ThorImageTaskStatus.Failed)
        {
            logger.TaskCompletedAt = DateTime.Now;
        }

        logger.UpdatedAt = DateTime.Now;

        await LoggerDbContext.SaveChangesAsync();

        // 发布任务状态更新事件
        await eventBus.PublishAsync(logger);
    }

    /// <summary>
    /// 根据TaskId获取任务日志
    /// </summary>
    public async ValueTask<ImageTaskLogger?> GetByTaskIdAsync(string taskId)
    {
        return await LoggerDbContext.ImageTaskLoggers
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.TaskId == taskId);
    }

    /// <summary>
    /// 分页查询图片任务日志
    /// </summary>
    public async ValueTask<PagingDto<ImageTaskLogger>> GetAsync(
        int page,
        int pageSize,
        ThorImageTaskType? taskType = null,
        ThorImageTaskStatus? taskStatus = null,
        string? model = null,
        DateTime? startTime = null,
        DateTime? endTime = null,
        string? keyword = null,
        string? userId = null)
    {
        var query = LoggerDbContext.ImageTaskLoggers.AsNoTracking();

        if (taskType.HasValue)
            query = query.Where(x => x.TaskType == taskType);

        if (taskStatus.HasValue)
            query = query.Where(x => x.TaskStatus == taskStatus);

        if (!string.IsNullOrWhiteSpace(model))
            query = query.Where(x => x.ModelName == model);

        if (startTime.HasValue)
            query = query.Where(x => x.CreatedAt >= startTime);

        if (endTime.HasValue)
            query = query.Where(x => x.CreatedAt <= endTime);

        // 用户权限校验
        if (!UserContext.IsAdmin)
        {
            query = query.Where(x => x.UserId == UserContext.CurrentUserId);
        }
        else if (!string.IsNullOrWhiteSpace(userId))
        {
            query = query.Where(x => x.UserId == userId);
        }

        if (!string.IsNullOrWhiteSpace(keyword))
        {
            query = query.Where(x =>
                x.UserName!.Contains(keyword) ||
                x.Prompt.Contains(keyword) ||
                x.TaskId.Contains(keyword) ||
                x.TokenName!.Contains(keyword) ||
                (!string.IsNullOrEmpty(x.ChannelName) && x.ChannelName.Contains(keyword)) ||
                x.ModelName.Contains(keyword)
            );
        }

        var total = await query.CountAsync();

        if (total <= 0)
            return new PagingDto<ImageTaskLogger>(total, []);

        var result = await query
            .OrderByDescending(x => x.CreatedAt)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        // 脱敏处理
        if (!UserContext.IsAdmin)
        {
            result.ForEach(x => { x.ChannelName = null; });
        }

        result.ForEach(x =>
        {
            if (!string.IsNullOrEmpty(x.TokenName))
            {
                x.TokenName = x.TokenName[..3] + "..." + x.TokenName[^3..];
            }
        });

        return new PagingDto<ImageTaskLogger>(total, result);
    }

    /// <summary>
    /// 获取任务统计信息
    /// </summary>
    public async ValueTask<object> GetTaskStatisticsAsync(DateTime? startTime = null, DateTime? endTime = null)
    {
        var query = LoggerDbContext.ImageTaskLoggers.AsNoTracking();

        if (startTime.HasValue)
            query = query.Where(x => x.CreatedAt >= startTime);

        if (endTime.HasValue)
            query = query.Where(x => x.CreatedAt <= endTime);

        if (!UserContext.IsAdmin)
        {
            query = query.Where(x => x.UserId == UserContext.CurrentUserId);
        }

        var statistics = await query
            .GroupBy(x => x.TaskType)
            .Select(g => new
            {
                TaskType = g.Key.ToString(),
                Total = g.Count(),
                Completed = g.Count(x => x.TaskStatus == ThorImageTaskStatus.Completed),
                Failed = g.Count(x => x.TaskStatus == ThorImageTaskStatus.Failed),
                Processing = g.Count(x => x.TaskStatus == ThorImageTaskStatus.Processing),
                TotalQuota = g.Sum(x => x.Quota)
            })
            .ToListAsync();

        return new
        {
            TaskTypes = statistics,
            TotalTasks = statistics.Sum(x => x.Total),
            TotalQuota = statistics.Sum(x => x.TotalQuota),
            SuccessRate = statistics.Sum(x => x.Total) > 0
                ? Math.Round((double)statistics.Sum(x => x.Completed) / statistics.Sum(x => x.Total) * 100, 2)
                : 0
        };
    }

    /// <summary>
    /// 根据渠道和TaskId列表查询任务
    /// 用于MJ的GetTaskListByCondition接口
    /// </summary>
    public async ValueTask<List<ImageTaskLogger>> GetTasksByChannelAndIdsAsync(
        string channelId,
        List<string> taskIds)
    {
        return await LoggerDbContext.ImageTaskLoggers
            .AsNoTracking()
            .Where(x => x.ChannelId == channelId && taskIds.Contains(x.TaskId))
            .ToListAsync();
    }
}